File: algorithms/[494]目标和/solution.py
# topics = ["动态规划"]

from typing import List


class Solution:
    """
    0-1 背包问题 变种问题
    """

    def findTargetSumWays(self, nums: List[int], S: int) -> int:
        total = sum(nums)

        if abs(S) > total:
            return 0

        n = len(nums)
        # dp[i][j] 表示数组区间 [0, i] 的数能否组成目标和 j - total 的方案数
        dp: List[List[int]] = [[0] * (2 * total + 1) for _ in range(n)]

        for i, num in enumerate(nums):
            for j in range(2 * total + 1):
                if i == 0:
                    dp[i][j] += num == j - total
                    dp[i][j] += num == total - j
                else:
                    if 0 <= j + num <= 2 * total:
                        dp[i][j] += dp[i - 1][j + num]
                    if 0 <= j - num <= 2 * total:
                        dp[i][j] += dp[i - 1][j - num]

        return dp[n - 1][S + total]

File: algorithms/[494]目标和/solution.ts
// topics = ["动态规划"]

function findTargetSumWays(nums: number[], S: number): number {
  const n = nums.length;
  const dp: number[][] = [];
  const sum = nums.reduce((acc, cur) => acc + cur, 0);

  for (let i = 0; i < n; i += 1) {
    dp[i] = [];
    for (let j = -sum; j <= sum; j += 1) {
      if (i == 0) {
        dp[i][j] = Number(nums[i] == j) + Number(-nums[i] == j);
      } else {
        dp[i][j] = (dp[i - 1][j - nums[i]] ?? 0) + (dp[i - 1][j + nums[i]] ?? 0);
      }
    }
  }

  return dp[n - 1][S] ?? 0;
}

function findTargetSumWays2(nums: number[], S: number): number {
  const n = nums.length;
  const sum = nums.reduce((acc, cur) => acc + cur, 0);
  let dp: number[] = [];

  for (let i = 0; i < n; i += 1) {
    let next: number[] = [];
    for (let j = -sum; j <= sum; j += 1) {
      if (i == 0) {
        next[j] = Number(nums[i] == j) + Number(-nums[i] == j);
      } else {
        next[j] = (dp[j - nums[i]] ?? 0) + (dp[j + nums[i]] ?? 0);
      }
    }
    dp = next;
  }

  return dp[S] ?? 0;
}
